src/vs/platform/telemetry/node/1dsAppender.ts

```diff
@@ -0,0 +1,158 @@
+						...payload.headers,
+						'Content-Type': 'application/json',
+						'Content-Length': Buffer.byteLength(payload.data)
+					}
+				};
+				try {
+					const req = https.request(payload.urlString, options, res => {
+						res.on('data', function (responseData) {
+							oncomplete(res.statusCode ?? 200, res.headers as Record<string, any>, responseData.toString());
+						});
+						// On response with error send status of 0 and a blank response to oncomplete so we can retry events
+						res.on('error', function (err) {
+							oncomplete(0, {});
+						});
+					});
+					req.write(payload.data);
+					req.end();
+				} catch {
+					// If it errors out, send status of 0 and a blank response to oncomplete so we can retry events
+					oncomplete(0, {});
+				}
+			}
+		};
+		// Configure the channel to use a XHR Request override since it's not available in node
+		const channelConfig: IChannelConfiguration = {
+			alwaysUseXhrOverride: true,
+			httpXHROverride: customHttpXHROverride
+		};
+		coreConfig.extensionConfig[collectorChannelPlugin.identifier] = channelConfig;
+	}
+
+	appInsightsCore.initialize(coreConfig, []);
+
+	appInsightsCore.addTelemetryInitializer((envelope) => {
+		if (envelope.tags) {
+			// Sets it to be internal only
+			envelope.tags['utc.flags'] = 880971612160;
+		}
+	});
+
+	return appInsightsCore;
+}
+
+
+export class OneDsAppender implements ITelemetryAppender {
+
+	private _aiCore: AppInsightsCore | undefined;
+	private _iKey: string | undefined;
+	private _asyncAiCore: Promise<AppInsightsCore> | null;
+
+	constructor(
+		private _eventPrefix: string,
+		private _defaultData: { [key: string]: any } | null,
+		iKeyOrClientFactory: string | (() => AppInsightsCore), // allow factory function for testing
+	) {
+		if (!this._defaultData) {
+			this._defaultData = Object.create(null);
+		}
+
+		if (typeof iKeyOrClientFactory === 'function') {
+			this._aiCore = iKeyOrClientFactory();
+		} else {
+			this._iKey = iKeyOrClientFactory;
+		}
+		this._asyncAiCore = null;
+	}
+
+	private _withAIClient(callback: (aiCore: AppInsightsCore) => void): void {
+		if (!this._aiCore) {
+			return;
+		}
+
+		if (this._aiCore) {
+			callback(this._aiCore);
+			return;
+		}
+
+		if (this._iKey && !this._asyncAiCore) {
+			this._asyncAiCore = getClient(this._iKey);
+			this._iKey = undefined;
+		}
+
+		this._asyncAiCore?.then(
+			(aiClient) => {
+				callback(aiClient);
+			},
+			(err) => {
+				onUnexpectedError(err);
+				console.error(err);
+			}
+		);
+	}
+
+	log(eventName: string, data?: any): void {
+		if (!this._aiCore) {
+			return;
+		}
+		data = mixin(data, this._defaultData);
+		data = validateTelemetryData(data);
+
+		// Attemps to suppress https://github.com/microsoft/vscode/issues/140624
+		try {
+			this._withAIClient((aiClient) => aiClient.track({
+				name: this._eventPrefix + '/' + eventName,
+				data: { ...data.properties, ...data.measurements },
+
+			}));
+		} catch { }
+	}
+
+	flush(): Promise<any> {
```

Comment:
So I assume that this method only gets called during "shutdown"?

Additionally, both AI and 1DS listen to the normal unload events and cause an automatic flush() to make sure that all batched events are sent. So generally, unless you are dynamically adding / destroying telemetry clients you (generally) don't need to call "unload" during shutdown unless your environment doesn't emit one of the unload DOM events ('unload', 'pagehide', 'beforeunload', 'visibilitychange') -- (so this doesn't include node environments -- you will still need to flush for there -- and I'm not sure if it has unload)

Reply:
This happens automatically in the PostChannel for 1DS and is part of the general main functionality for AI.

Reply:
So I'm not saying that you should not do this, just that you may not need to.

Reply:
Sometimes we manually destroy the client, so I will just leave it since we know this works and I figured it can't hurt :)